openlibrary/plugins/upstream/account.py

```diff
@@ -1219,35 +1219,41 @@ def get_loan_history_data(page: int, mb: "MyBooksTemplate") -> dict[str, str | i
         loan_history.pop()
 
     ocaids = [loan_record['identifier'] for loan_record in loan_history]
-    availability = get_availability_of_ocaids(ocaids)
+    ocaid_availability = get_availability_of_ocaids(ocaids)
     loan_history_map = {
         loan_record['identifier']: loan_record for loan_record in loan_history
     }
     editions = web.ctx.site.get_many(
         [
-            '/books/%s' % availability[ocaid].get('openlibrary_edition')
-            for ocaid in availability
-            if availability[ocaid].get('openlibrary_edition')
+            f"/books/{item.get('openlibrary_edition')}"
+            for item in ocaid_availability.values()
+            if item.get('openlibrary_edition')
         ]
     )
 
     # Create 'placeholder' editions for items in the Internet Archive loan
     # history, but absent from Open Library.
     ia_only_loans = [
-        loan for loan in availability.values() if not loan.get('openlibrary_edition')
+        loan
+        for loan in ocaid_availability.values()
+        if not loan.get('openlibrary_edition')
     ]
 
-    # Attach availability and loan to both editions and ia-only items.
+    # Attach availability and loan history info (for sorting) to both editions and
+    # ia-only items.
     for ed in editions:
         if ed.ocaid in ocaids:
-            ed.availability = availability.get(ed.ocaid)
-            ed.loan = loan_history_map[ed.ocaid]
```

Comment:
This was presumably causing a problem because in LoanStatus, we check a book method called ed.loan to fetch any loan that may be attached to the book and what we're doing here (which is attaching a loan **history** record to the book) was technically conflating two different loan types.

https://github.com/internetarchive/openlibrary/blob/master/openlibrary/macros/LoanStatus.html#L37-L46C38

The possible solutions are either to:
1. not attach the loan to the edition (if it's not being used)
2. attach the loan to the edition under a different name like `loan_history`
3. if the `loan_history` record has enough info associated to determine if it's active (e.g. an expiration date) and also assuming the ed.loan_history is in the same format as ed.loan else where on openlibrary.org. then consider modifying LoanStatus.html's check ~L37 to consider the expiry when determining if a book should show up as read.